Removing requiring dir attribute to be on a <span> element in language plugin

KarolDawidziuk opened the pull request:
<!--
🚨 If you want to submit a PR for a security vulnerability, please contact us directly
at https://ckeditor.com/contact/ instead. 🚨
-->
## What is the purpose of this pull request?

Feature

## Does your PR contain necessary tests?

All patches that change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [x] Manual tests

## Did you follow the CKEditor 4 code style guide?

Your code should follow the guidelines from the [CKEditor 4 code style guide](https://github.com/ckeditor/ckeditor4/blob/major/dev/docs/codestyle.md) which helps keep the entire codebase consistent.

- [x] PR is consistent with the code style guide

## What is the proposed changelog entry for this pull request?

```
* [#5085](https://github.com/ckeditor/ckeditor4/issues/5085): Fixed: The [language](https://ckeditor.com/cke4/addon/language) plugin removed the `span` element when it did not have a `dir` attribute.
```

## What changes did you make?

Language plugin required `dir` attribute on a span element and element without this attribute was removed. 
Fix finds all `span` elements with `lang` and without the `dir` attribute, or with an empty `dir` attribute. Then, based on the value in the `lang` attribute, the correct value for the `dir` attribute is determined.

## Which issues does your PR resolve?

Closes #5085.

Comandeer commented on tests/plugins/language/manual/addmissingdirattribute.md:
```diff
@@ -0,0 +1,14 @@
+@bender-tags: 4.19.1, feature, 5085
```
```suggestion
@bender-tags: 4.20.0, feature, 5085
```

jacekbogdanski commented:
I changed the original issue to a bug, so we can target that PR into `master` branch.

KarolDawidziuk commented:
Thanks, @Comandeer for the review.

I applied the changes mentioned in your comment and PR is ready for another review.
Also, I resolved comments about changing bender tags, as we decided to point this PR to the `master` branch instead of `next`.

Comandeer commented:
Rebased onto latest `master`.

Comandeer commented on plugins/language/plugin.js:
```diff
@@ -47,6 +49,27 @@
 			} );
 
+			// Add missing proper `dir` attribute when <span> element has only `lang` attribute (#5085).
+			editor.on( 'toHtml', function( evt ) {
+				var langElements = evt.data.dataValue.find( function( child ) {
+					return child.name === 'span' &&
+						child.attributes.lang && ( !child.attributes.dir || child.attributes.dir === '' );
+				}, true );
+
+				if ( langElements.length === 0 ) {
+					return;
+				}
+
+				CKEDITOR.tools.array.forEach( langElements, function( element ) {
+					var rtlLanguages = CKEDITOR.tools.object.keys( CKEDITOR.lang.rtl ),
+						isRtlLanguage = CKEDITOR.tools.array.indexOf( rtlLanguages, element.attributes.lang ) !== -1,
+						dirAttribute = isRtlLanguage ? 'rtl' : 'ltr';
+
+					element.attributes.dir = dirAttribute;
+				} );
+
+			}, null, null, 10 );
```
I was quite sure that we have some kind of API dedicated for such use cases. And it seems that we, indeed, have one! Although it's undocumented :confused: (see #4839, #5196, or #4585). Yet it can be reverse-engineered through the code of our other plugins, e.g. [`colorbutton`](https://github.com/ckeditor/ckeditor4/blob/3dec7a87364e14c6e40791575c7a537109a5eb05/plugins/colorbutton/plugin.js#L35-L51). Buttons added to the toolbar via the `editor.ui#add()` method can have a `contentTransformations` property that can – as its name suggests – transform content input to the editor. The syntax of transformations is described in detail in [the `editor.filter#addTransformations()` method's docs](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_filter.html#method-addTransformations).

KarolDawidziuk replied:
I applied `contentTransformations` and it looks much better. Thank you. 😉

Comandeer commented on tests/plugins/language/language.js:
```diff
@@ -144,6 +144,123 @@
+						resume( function() {
+							var expected = '';
+
+							if ( CKEDITOR.env.gecko || ( CKEDITOR.env.ie && CKEDITOR.env.version < 10 ) ) {
+								expected = '<p>foo <span dir="rtl" lang="' + language + '">bar</span></p>';
+							} else {
+								expected = '<p>foo <span lang="' + language + '" dir="rtl">bar</span></p>';
+							}
+
+							assert.areSame(
+								editor.editable().getData(), expected,
+								'The value of dir attribute for ' + language + ' language is incorrect.'
+							);
```
Can't it be simplified by using `assert.beautified.html()`? It has an option to sort attributes, so we won't need two distinctive versions with all the same but attributes order

KarolDawidziuk replied:
Yes, it can be used here, good point. 👍  Tests are already updated.

Comandeer commented on tests/plugins/language/language.js:
```diff
@@ -144,6 +144,123 @@
+						resume( function() {
+							var expected = '';
+
+							if ( CKEDITOR.env.gecko || ( CKEDITOR.env.ie && CKEDITOR.env.version < 10 ) ) {
+								expected = '<p>foo <span dir="ltr" lang="' + language + '">bar</span></p>';
+							} else {
+								expected = '<p>foo <span lang="' + language + '" dir="ltr">bar</span></p>';
+							}
+
+							assert.areSame(
+								editor.editable().getData(), expected,
+								'The value of dir attribute for ' + language + ' language is incorrect.'
+							);
```
The same here.

Comandeer commented on tests/plugins/language/language.js:
```diff
@@ -144,6 +144,123 @@
+						resume( function() {
+							var expected = '';
+
+							if ( CKEDITOR.env.gecko || ( CKEDITOR.env.ie && CKEDITOR.env.version < 10 ) ) {
+								expected = '<p>foo <span dir="' + dir + '" lang="' + language + '">bar</span></p>';
+							} else {
+								expected = '<p>foo <span lang="' + language + '" dir="' + dir + '">bar</span></p>';
+							}
+
+							assert.areSame(
+								editor.editable().getData(), expected,
+								'The value of dir attribute for ' + language + ' language is incorrect.'
+							);
```
And here.

Comandeer reviewed (changes requested):
The code after the refactor looks really nice, good job! But I've made some research and was able to find an API that was designed for such use cases. Could you take a look?

KarolDawidziuk commented:
Thanks @Comandeer for the review.

I've moved the fix logic to the dedicated `contentTransformations` API, updated unit tests and PR is ready for another review.

Comandeer reviewed (changes requested):
Looks really good 👍🏻 !

Yet, it seems that `contentTransformations` has some quirk: they are not applied if the `allowedContent` is set to true (so basically: if the filter is off in the editor). And due to that, `span[lang]:not([dir])` are not recognized as language elements by the plugin. It's probably caused by the logic inside the `getCurrentLangElement()` method (it checks for both `[lang]` and `[dir]` attributes). As we already updated `requiredContent` for the plugin, we should also update the logic inside `getCurrentLangElement()` – WDYT?

KarolDawidziuk commented:
Thanks, @Comandeer for the review.

I updated `getCurrentLangElement()` method, added manual and unit test with removed filters to check if the language plugin correctly detects language elements and PR is ready for another check.

Comandeer reviewed (approved):
LGTM! I've just slightly reworded the names of newly added unit tests and changed the `feature` tag to the `bug` one in manual tests.
